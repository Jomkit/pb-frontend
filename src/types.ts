interface IProject {
    id: string,
    name: string,
    note: string
}

interface ITask {
    id: string,
    projectId?: string,
    name: string,
    duration?: string,
    estimate?: string,
    status?: string
}

// Not sure this one is needed yet
interface ISurvey {
    id?: string,
    projectId: string,
    taskId: string,
    timerId: string,
    score: number,
    description?: string
}

interface IUser {
    id?: number,
    username: string,
    password?: string,
    firstName?: string,
    lastName?: string,
    email?: string
}

export type { IProject, ITask, ISurvey, IUser }