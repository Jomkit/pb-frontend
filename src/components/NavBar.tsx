import React, { useContext } from "react";
import { Link, useNavigate } from "react-router-dom"
import userContext from "./contexts/userContext";

const NavBar = ({ currUserToken, logout }: { currUserToken?: string, logout: () => void }) => {
  const currUser = useContext(userContext);

  const navigate = useNavigate();

  const handleLogout = (evt: React.MouseEvent<HTMLAnchorElement>) => {
    evt.preventDefault();
    logout();
    navigate("/login");
  }
  
  return (
    <nav className='flex justify-between items-center mx-5'>
        <div className='text-lg space-x-3'>
          <Link className="text-3xl" to="/">PB</Link> 
          {currUserToken ?
          <>
            <Link to="/tasks">Tasks</Link>
            <Link to="/projects">Projects</Link>
          </>
          :
          <></>
          }
        </div>
        <div className='text-lg space-x-3'>
          {currUserToken ? 
          <>
            <Link to="/profile">{currUser.username}</Link>
            <Link to="#" onClick={handleLogout}>Logout</Link>
          </>
          :
          <>
            <Link to="/login">Login</Link>
            <Link to="/signup">Sign Up</Link>
          </>
          }
        </div>
    </nav>
  )
}

export default NavBar