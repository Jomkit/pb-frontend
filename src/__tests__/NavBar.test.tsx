import { render } from "@testing-library/react"
import NavBar from "../components/NavBar";
import UserContext from "../components/contexts/userContext";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { IUser } from "../types";

describe("NavBar Component", function(){ 
    const mockedLogout = vi.fn();    
    
    it('should render', () => { 
        render(
            <MemoryRouter>
                <NavBar logout={mockedLogout} />
            </MemoryRouter>
    );
    })

    it('should match snapshot of no logged-in user', () => { 
        const { asFragment } = render(
            <MemoryRouter>
                <NavBar logout={mockedLogout} />
            </MemoryRouter>
        )
        expect(asFragment()).toMatchSnapshot();
    })

    it('should match snapshot of logged-in user', () => {
        const testUser:IUser = {
            username: "testUser", 
            id: 1,
            firstName: "test",
            lastName: "user",
            email: "test@user"
        }
        
        const { asFragment, getByText } = render(
            <MemoryRouter>
                <UserContext.Provider value={testUser}>
                    <NavBar currUserToken="fakeToken" logout={mockedLogout} />
                </UserContext.Provider>
            </MemoryRouter>
        )
        expect(asFragment()).toMatchSnapshot();
        expect(getByText("testUser")).toBeInTheDocument();
        expect(getByText("Logout")).toBeInTheDocument();
    })

    it('should call logout', async () => { 
        const testUser:IUser = {
            username: "testUser", 
            id: 1,
            firstName: "test",
            lastName: "user",
            email: "test@user"
        }

        const user = userEvent.setup();

        const { getByText } = render(
            <MemoryRouter>
                <UserContext.Provider value={testUser}>
                    <NavBar currUserToken="fakeToken" logout={mockedLogout} />
                </UserContext.Provider>
            </MemoryRouter>
        )

        const logoutBtn = getByText("Logout");
        await user.click(logoutBtn);
        expect(mockedLogout).toHaveBeenCalled();
    })
})