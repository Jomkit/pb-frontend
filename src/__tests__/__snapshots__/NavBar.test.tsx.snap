// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavBar Component > should match snapshot of logged-in user 1`] = `
<DocumentFragment>
  <nav
    class="flex justify-between items-center mx-5"
  >
    <div
      class="text-lg space-x-3"
    >
      <a
        class="text-3xl"
        href="/"
      >
        PB
      </a>
      <a
        href="/tasks"
      >
        Tasks
      </a>
      <a
        href="/projects"
      >
        Projects
      </a>
    </div>
    <div
      class="text-lg space-x-3"
    >
      <a
        href="/profile"
      >
        testUser
      </a>
      <a
        href="/"
      >
        Logout
      </a>
    </div>
  </nav>
</DocumentFragment>
`;

exports[`NavBar Component > should match snapshot of no logged-in user 1`] = `
<DocumentFragment>
  <nav
    class="flex justify-between items-center mx-5"
  >
    <div
      class="text-lg space-x-3"
    >
      <a
        class="text-3xl"
        href="/"
      >
        PB
      </a>
    </div>
    <div
      class="text-lg space-x-3"
    >
      <a
        href="/login"
      >
        Login
      </a>
      <a
        href="/signup"
      >
        Sign Up
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
